server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/ImageUploadFieldWidget.java

```diff
@@ -0,0 +1,328 @@
+            oldHandler.removeHandler();
+        }
+
+        oldHandler = formPanel.addSubmitCompleteHandler(new FormPanel.SubmitCompleteHandler() {
+            @Override
+            public void onSubmitComplete(FormPanel.SubmitCompleteEvent event) {
+                // event.getResults is always null because of cross-domain upload
+                // we are forced to make additional call to check whether upload is successful
+
+                fetchImageServingUrl();
+            }
+        });
+        formPanel.submit();
+    }
+
+    public void fetchImageServingUrl() {
+        try {
+            RequestBuilder requestBuilder = new RequestBuilder(RequestBuilder.GET, uploader.getBaseUrl() + "/imageUrl");
+            requestBuilder.sendRequest(null, new RequestCallback() {
+                @Override
+                public void onResponseReceived(Request request, Response response) {
+                    servingUrl = response.getText();
+                    setStateAfterUpload();
+                }
+
+                @Override
+                public void onError(Request request, Throwable exception) {
+                    servingUrl = null;
+                    Log.error("Failed to fetch image serving url. ", exception);
+                    setStateAfterUpload();
+                }
+            });
+        } catch (Exception e) {
+            servingUrl = null;
+            Log.error("Failed to send request for fetching serving url. ", e);
+            setStateAfterUpload();
+        }
+    }
+
+    public boolean isValid() {
+        return !Strings.isNullOrEmpty(servingUrl);
+    }
+
+    private void setStateAfterUpload() {
+        setLoadingState(false);
+
+        boolean valid = isValid();
+        downloadButton.setVisible(valid);
+        clearButton.setVisible(valid);
+        image.setUrl(valid ? servingUrl + "=s" + formPanel.getOffsetWidth() : "");
+        uploadFailed.setVisible(!valid);
+
+        if (valid) {
+            valueUpdater.update(getValue());
+        } else {
+            Log.error("Failed to fetch image serving url.");
+        }
+    }
+
+    private AttachmentValue getValue() {
+        AttachmentValue value = new AttachmentValue();
+        value.getValues().add(attachment.get());
+        return value;
+    }
+
+    @Override
+    public void setReadOnly(boolean readOnly) {
+        this.readOnly = readOnly;
+    }
+
+    @Override
+    public Promise<Void> setValue(AttachmentValue value) {
+        clearValue();
+
+        if (value != null && value.getValues() != null && value.getValues().size() > 0) {
+            attachment.set(value.getValues().iterator().next());
+            fetchImageServingUrl();
+        }
+
+        return Promise.done();
+    }
+
+    @Override
+    public void setType(FieldType type) {
+        // ignore
+    }
+
+    @Override
+    public void clearValue() {
+        image.setUrl("");
+        servingUrl = null;
+        attachment.set(new Attachment());
```

Comment:
Why not just `attachment = Optional.absent()` or `Optional.of(new Attachment())` ?